components/script/dom/element.rs

```diff
@@ -3582,6 +3586,38 @@ impl Element {
         element
     }
 
+    pub fn is_invalid(&self, needs_update: bool) -> bool {
+        if let Some(validatable) = self.as_maybe_validatable() {
+            if needs_update {
+                validatable
+                    .validity_state()
+                    .perform_validation_and_update(ValidationFlags::all());
+            }
+            return !validatable.is_instance_validatable() || validatable.satisfies_constraints();
+        }
+
+        if let Some(internals) = self.get_element_internals() {
+            return internals.is_invalid();
+        }
+        true
+    }
```

Comment:
Because `validatable` including the form control elements and elements with internals, it is not easy to check both outside element.rs, so we create functions here.